pkg/controller/storagecluster/controller_test.go

```diff
@@ -1341,6 +1347,102 @@ func TestStoragePodGetsScheduledWithCustomNodeSpecs(t *testing.T) {
 	require.Equal(t, *clusterRef, podControl.ControllerRefs[2])
 }
 
+// When a StorageNode is unhealthy, and the storage node is not a k8s node. we should not update storage pod as
+// it may cause storage cluster to lose quorum.  This could happen if a user removes a k8s node from the cluster
+// for maintenance, and meanwhile tries to uppgrade portworx.
+func TestStoragePodUpdateWhenStorageNodeUnhealthy(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	driverName := "mock-driver"
+	cluster := createStorageCluster()
+	k8sVersion, _ := version.NewVersion(minSupportedK8sVersion)
+	driver := testutil.MockDriver(mockCtrl)
+	storageLabels := map[string]string{
+		constants.LabelKeyClusterName: cluster.Name,
+		constants.LabelKeyDriverName:  driverName,
+	}
+	k8sClient := testutil.FakeK8sClient(cluster)
+	podControl := &k8scontroller.FakePodControl{}
+	recorder := record.NewFakeRecorder(10)
+	controller := &Controller{
+		client:              k8sClient,
+		Driver:              driver,
+		podControl:          podControl,
+		recorder:            recorder,
+		kubernetesVersion:   k8sVersion,
+		lastPodCreationTime: make(map[string]time.Time),
+	}
+
+	var storageNodes []*storageapi.StorageNode
+	storageNodes = append(storageNodes, createStorageNode("k8s-node-1", true))
+	storageNodes = append(storageNodes, createStorageNode("k8s-node-2", true))
+	storageNodes = append(storageNodes, createStorageNode("k8s-node-3", true))
+	storageNodes = append(storageNodes, createStorageNode("not-k8s-node", false))
```

Comment:
This is a nice test!
Can we fit in the scenarios in the same or a different test:
- where a storage node is offline, but on a valid kubernetes node (we are only checking offline storage node on non-k8s node)
- a k8s node for which`wantToRun` comes as false, but storage node is present (online or offline)

Reply:
@piyush-nimbalkar  we actually check offline storage nodes on k8s nodes and non-k8s nodes, so if a storage node is offline and it is also a k8s node, we will also not proceed with upgrade. 

`we are only checking offline storage node on non-k8s node`